Address signature reuse vulnerability

bifurcation opened the pull request:
This PR addresses the signature reuse vulnerability noted by Andrew Ayer.

https://mailarchive.ietf.org/arch/msg/acme/F71iz6qq1o_QPVhJCV4dqWf-4Yc

It removes the unnecessary signatures on validation objects used in challenges, and replaces it with a simpler structure where the domain holder specifies which account keys are authorized, through which challenges.

Currently WIP, for discussion.

jdkasten commented:
Can you link or write out the specifics of the proposed solution?

It will be easier to reason about and discuss.

bifurcation commented:
I was waiting for @jsha to do that in a spec PR :)  Basically:

First: Define an "Authorized keys" object format:
```
[{
  "token": /* challenge.Token */
  "key": /* challenge.AccountKey */
}]
```

For SimpleHTTP:
* have the server provide that object instead of the signed validation object
* check that the token and key fields match those in the challenge.

For DVSNI and DNS:
* have the client's response include a serialized authorized keys object
* check that that corresponds to the challenge before sending the validation query
* have the validation query return a digest of the serialized object (either as `Zname` or the contenst of the TXT record)

bifurcation commented:
After some discussion with @ekr , I've switched this around a little bit.  In order to avoid attacks where a client manipulates the representation of the authorizedKeys object to obtain a given digest, the server now dictates the serialized authorized keys object to the client.  In effect, the server is telling the client what digest to provision, but in a way that the client can verify that the value has the intended semantic.

r0ro commented on va/validation-authority.go:
```diff
@@ -523,7 +499,11 @@ func (va *ValidationAuthorityImpl) validateDNS(identifier core.AcmeIdentifier, i
+	// Compute the digest of the authorized keys file
+	h := sha256.New()
+	h.Write([]byte(challenge.AuthorizedKeys))
+	authorizedKeysDigest := hex.EncodeToString(h.Sum(nil))
```
I like the fact that a hash is used here, this will simplify the dns validation implementation by not requiring to store long txt records.

r0ro replied:
found the related acme PR https://github.com/letsencrypt/acme-spec/pull/223

kuba commented:
I guess this should be pulled only when https://github.com/letsencrypt/acme-spec/pull/223 goes through. I would also ask you not to merge before we get this done on the client side - otherwise we'll both suffer from integration issues.

bifurcation commented:
@kuba Note that this PR disables the le-client integration tests, for exactly this reason.

bifurcation commented on core/objects_test.go:
```diff
@@ -40,6 +42,29 @@ func TestRegistrationUpdate(t *testing.T) {
 
+var testKey1, _ = rsa.GenerateKey(rand.Reader, 2048)
+var testKey2, _ = rsa.GenerateKey(rand.Reader, 2048)
+var testKey3, _ = rsa.GenerateKey(rand.Reader, 2048)
```
This key is no longer used.  Delete.

bifurcation commented on ra/registration-authority_test.go:
```diff
@@ -431,7 +437,11 @@ func TestUpdateAuthorizationReject(t *testing.T) {
+	authzUpdated, err := sa.NewPendingAuthorization(AuthzInitial)
+	test.AssertNotError(t, err, "Failed to create new pending authz")
+	fmt.Println("AUTHZ_UPDATED", authzUpdated)
+	fmt.Println("AUTHZ_UPDATED_ERR", err)
```
Remove `fmt.Println` statements here.

bifurcation commented on test/amqp-integration-test.py:
```diff
@@ -129,6 +129,7 @@ def cleanup():
 if not startservers.start(race_detection=True):
     die(ExitStatus.Error)
 run_node_test()
+## TODO(rlb@ipv.sx) Re-enable Let's Encrypt client tests once the client is upgraded
```
If the LE client works now, delete this line.

bifurcation commented on test/js/test.js:
```diff
@@ -348,26 +347,18 @@ function getReadyToValidate(err, resp, body) {
+  var challenge = httpChallenges[0];
+  var jsonAuthorizedKey = util.b64dec(challenge.authorizedKey)
+  var authorizedKey = JSON.parse(jsonAuthorizedKey);
+  var challengePath = ".well-known/acme-challenge/" + authorizedKey.token;
```
Need to add a check here that the authorized key is the correct one.

bifurcation commented on test/js/test.js:
```diff
@@ -376,29 +367,29 @@ function getReadyToValidate(err, resp, body) {
         req.url == "/" + challengePath) {
-      response.writeHead(200, {"Content-Type": "application/jose+json"});
-      response.end(JSON.stringify(validationJWS));
+      console.log("\nRespodned with authorized key file:", jsonAuthorizedKey);
```
Remove `console.log` statements.

bifurcation commented on test/js/test.js:
```diff
@@ -376,29 +367,29 @@ function getReadyToValidate(err, resp, body) {
+    console.log("listening on port 5001");
     state.httpServer.listen(5001)
   } else {
+    console.log("listening on port 443");
```
Here as well.

bifurcation commented on va/validation-authority.go:
```diff
@@ -330,18 +282,142 @@ func (va *ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdenti
+	return body, challenge, nil
+}
+
+func (va *ValidationAuthorityImpl) validateZName(identifier core.AcmeIdentifier, input core.Challenge, zName string) (core.Challenge, error) {
```
A better name for this function would be welcome.

bifurcation commented on va/validation-authority.go:
```diff
@@ -330,18 +282,142 @@ func (va *ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdenti
+}
+
+//-----BEGIN TO DELETE-----
+func verifyValidationJWS(validation *jose.JsonWebSignature, accountKey *jose.JsonWebKey, target map[string]interface{}) error {
```
It will clean up the diff if this were moved back up top. Just add separate "TO DELETE" fencing for it.

bifurcation commented on va/validation-authority.go:
```diff
@@ -417,67 +493,138 @@ func (va *ValidationAuthorityImpl) validateDvsni(identifier core.AcmeIdentifier,
-	}
-	if problem != nil {
+	return va.validateZName(identifier, challenge, ZName)
+}
```
`-----END TO DELETE-----` ?

bifurcation commented on va/validation-authority.go:
```diff
@@ -417,67 +493,138 @@ func (va *ValidationAuthorityImpl) validateDvsni(identifier core.AcmeIdentifier,
+	return va.validateZName(identifier, challenge, ZName)
+}
+
+func (va *ValidationAuthorityImpl) validateHTTP01(identifier core.AcmeIdentifier, input core.Challenge) (core.Challenge, error) {
```
This should go outside of the `TO DELETE` fencing.

bifurcation commented on va/validation-authority.go:
```diff
@@ -532,7 +675,23 @@ func (va *ValidationAuthorityImpl) validateDNS(identifier core.AcmeIdentifier, i
-	encodedSignature := core.B64enc(challenge.Validation.Signatures[0].Signature)
+
+	// Check that the account key for this challenge is authorized by this object
+	if !authorizedKey.Match(challenge.Token, challenge.AccountKey) {
```
Rename this method to `validateDNS01`

bifurcation commented on va/validation-authority_test.go:
```diff
@@ -539,18 +519,412 @@ func TestTLSError(t *testing.T) {
+	fmt.Println(finChall)
+}
+
+func getPort(hs *httptest.Server) (int, error) {
```
It would clean up the diff to move this back to its old position.

bifurcation commented on va/validation-authority_test.go:
```diff
@@ -539,18 +519,412 @@ func TestTLSError(t *testing.T) {
+	return hs
+}
+
+func brokenTLSSrv() *httptest.Server {
```
It would clean up the diff to move this back to its original position.

bifurcation commented:
OK, I think this is finally ready for review.  This PR now implements the latest proposed changes to the spec:

* Adds challenges `http-01` and `tls-sni-01` that implement the [updated, digest-based challenges](https://github.com/ietf-wg-acme/acme/pull/6) and [protections against default vhosts for the HTTP challenge](https://github.com/ietf-wg-acme/acme/pull/7)
* Changes the `dns` challenge to be `dns-01` for consistency with the [new naming scheme](https://github.com/ietf-wg-acme/acme/pull/8)  (There's no compatibility risk here: Since boulder doesn't offer the DNS challenge, nobody is using it who's not running their own stack.)

jcjones commented on core/objects.go:
```diff
@@ -239,6 +264,21 @@ type Challenge struct {
+// UnsafeSetToken sets the token value both in the Token field and in the
+// serialized AuthorizedKey object.
+//
+// This method should only be used for writing tests, in cases where the
```
I poked around a little to see if there was something we could do to try and catch use of this method in prod code but I don't see much other than conditional compilation.

Anyway, I'm sure it's fine. Just was curious.

jmhodges replied:
We could put it in test and make it UnsafeSetToken(ch *core.Challenge, token string)?

bifurcation replied:
Good idea.  I'll do that.

jmhodges replied:
This doesn't seem to have happened.

jcjones commented on ra/registration-authority_test.go:
```diff
@@ -431,7 +437,9 @@ func TestUpdateAuthorizationReject(t *testing.T) {
 	defer cleanUp()
-	authzUpdated, _ := sa.NewPendingAuthorization(AuthzInitial)
+	authzUpdated, err := sa.NewPendingAuthorization(AuthzInitial)
+	test.AssertNotError(t, err, "Failed to create new pending authz")
```
good catch

jcjones commented on va/validation-authority.go:
```diff
@@ -417,67 +496,138 @@ func (va *ValidationAuthorityImpl) validateDvsni(identifier core.AcmeIdentifier,
+	// Check that the account key for this challenge is authorized by this object
+	if !serverAuthorizedKey.MatchAuthorizedKey(authorizedKey) {
+		err = fmt.Errorf("The authorizated keys file from the server did not match this challenge")
+		va.log.Debug(err.Error())
```
This kind of logging is going to be hard to diagnose in staging/production, as these log entries may not arrive adjacently depending on load.

I would recommend these get concatenated together, perhaps using the DebugObject method.

bifurcation replied:
I think these lines were a testing artefact.  I've changed the content of `err` to have the JSON form of the two authorized key objects.

jmhodges commented on core/challenges.go:
```diff
@@ -5,31 +5,73 @@
+	}, nil
+}
+
+//----- BEGIN TO DELETE -----
```
Remove this and its pair and instead drop some links to issue tickets to remove them in these functions godoc. The tickets can be more clear as to when and why we'd delete them.

These comments as is will make people who don't know about this change nervy in the future when they come across it. "Should we delete this now? How did it get in??" that kind of thing.

bifurcation replied:
Created #894 for this, and updated the comments to point to that issue.

jmhodges replied:
Which comments did you update? Sorry, not seeing them in this draft.

jmhodges commented:
Sorry, early punch of buttons. Not done reviewing.

jcjones commented:
r=jcjones with my logging thing addressed, and preferably also the comment lines @jmhodges caught above.

I'll re-review quickly if anything else needs review.

jmhodges commented on core/objects.go:
```diff
@@ -191,6 +193,26 @@ type ValidationRecord struct {
+	if ak.Key == nil {
+		return false
+	}
+	return token == ak.Token && KeyDigestEquals(key.Key, ak.Key.Key)
```
I think these need to be a crypto/subtle.ConstantTimeCompare comparison?

jmhodges replied:
Please respond to my comment about why crypto/subtle is not needed here or make the changes.

jmhodges commented on core/objects.go:
```diff
@@ -265,16 +309,28 @@ func (ch Challenge) RecordsSane() bool {
+
+// LegacyIsSane performs sanity checks for legacy challenge types, which have
+// a different structure / logic than current challenges.
+func (ch Challenge) LegacyIsSane(completed bool) bool {
```
this doesn't need to be public. when it goes private, I'm not sure the isLegacy check below is as necessary.

jmhodges commented on core/objects.go:
```diff
@@ -265,16 +309,28 @@ func (ch Challenge) RecordsSane() bool {
+//----- BEGIN TO DELETE -----
+// IsLegacy returns true if the challenge is of a legacy type (i.e., one defined
+// before draft-ietf-acme-acme-00)
+func (ch Challenge) IsLegacy() bool {
```
this doesn't need to be public. make it isLegacy

jmhodges commented on core/objects.go:
```diff
@@ -329,9 +382,9 @@ func (ch Challenge) IsSane(completed bool) bool {
+// LegacyMergeResponse copies a subset of client-provided data to the current Challenge.
 // Note: This method does not update the challenge on the left side of the '.'
-func (ch Challenge) MergeResponse(resp Challenge) Challenge {
+func (ch Challenge) LegacyMergeResponse(resp Challenge) Challenge {
```
this doesn't need to be public. API consumers would just use MergeResponse

jmhodges commented on core/challenges.go:
```diff
@@ -5,31 +5,73 @@
+//----- BEGIN TO DELETE -----
 // SimpleHTTPChallenge constructs a random HTTP challenge
-func SimpleHTTPChallenge() Challenge {
+func SimpleHTTPChallenge(accountKey *jose.JsonWebKey) (Challenge, error) {
```
This doesn't ever return an error. Delete it from the return value

jmhodges replied:
You didn't remove these errors nor leave a comment on my thread saying why you didn't.

jmhodges commented on core/challenges.go:
```diff
@@ -5,31 +5,73 @@
 
 // DvsniChallenge constructs a random DVSNI challenge
-func DvsniChallenge() Challenge {
+func DvsniChallenge(accountKey *jose.JsonWebKey) (Challenge, error) {
```
this doesn't ever return an error so we don't need it in the return type

jmhodges commented on core/challenges.go:
```diff
@@ -5,31 +5,73 @@
+		Key:   accountKey,
+	}
+
+	jsonAK, err := json.Marshal(ak)
```
We marshal this ak three times in pa.ChallengesFor.

Why can't we wait to serialize the ak when we serialize the Challenge?

jmhodges replied:
You didn't respond to my comment about this marshalling the AK three times in ChallengesFor. I asked: why do we needed to serialize this here, instead of at the time the Challenge gets serialized?

jmhodges commented:
Also, I see some back and forth on the ietf ACME list about this patch. I know we don't want to block ourselves on the IETF always, but the back and forth includes the original reporter of the bug that this patch is for. It'd be cool for that to be resolved before we take this patch.

bifurcation commented:
@jcjones @jmhodges I think this PR now addresses your issues, and it's up-to-date with the [proposed spec](https://github.com/ietf-wg-acme/acme/pull/13).  Re-r?

jcjones commented on core/util.go:
```diff
@@ -144,6 +146,14 @@ func NewToken() string {
+// LooksLikeAToken checks whether a string represents a 32-octet value in
+// the URL-safe base64 alphabet.
+func LooksLikeAToken(token string) bool {
+	return regexp.MustCompile(tokenFormat).MatchString(token)
```
Since this gets run pretty often, it'd be a performance win to store the compiled regex and re-use it. That could be another issue/PR though.

jcjones commented on core/util.go:
```diff
@@ -154,6 +164,96 @@ func Fingerprint256(data []byte) string {
+}
+
+// Thumbprint computes the JWK Thumbprint of a key using the
+// indicated hash algorithm.
```
Nit: Not really the indicated hash, it's always SHA-256.

jcjones commented on core/util_test.go:
```diff
@@ -58,12 +64,22 @@ const JWK1JSON = `{
   "e":"AQAB"
 }`
 
+func TestKeyThumbprint(t *testing.T) {
```
This seems really light on the tests, but I'm willing to go with it if the upstream version is properly tested, since this is back-ported code.

bifurcation replied:
Yeah, the upstream version passes a more thorough test suite, and the code here is a pretty straight copy/paste.  I think this is tolerable until we get the upstream landed and merged.

jcjones commented on sa/_db/migrations/20151003141825_AddKeyAuthorization.sql:
```diff
@@ -0,0 +1,12 @@
+-- +goose Up
+-- SQL in section 'Up' is executed when this migration is applied
+
+ALTER TABLE `challenges` ADD COLUMN (`keyAuthorization` mediumblob);
```
Is this really bigger than `2^16` (the size of `blob`)? Does it need to support 16 MB/authorization?

bifurcation replied:
Good point.  In fact, `keyAuthorization` is just a string, and a well-formed one is always 86 octets long.  Just in case (and for consistency with other fields), I'll use `varchar(255)`.

jcjones commented on sa/storage-authority_test.go.orig:
```diff
@@ -0,0 +1,496 @@
+// Copyright 2014 ISRG.  All rights reserved
```
This looks like a merge error?

jcjones commented on ra/registration-authority_test.go.orig:
```diff
@@ -0,0 +1,714 @@
+// Copyright 2014 ISRG.  All rights reserved
```
Also maybe a merge error?

jcjones commented:
Check the `*.orig` files, and do comment on the 16 MB vs 64 kB column size; it sorta matters since it will affect how quickly the tables will stop holding that column in memory and force it to disk.

With the DB column commented on (If it needs to be mediumblob, that's OK!)  and merge things fixed, r=jcjones

jmhodges commented:
conflicts :/

bifurcation commented:
Merged!

kuba commented on cmd/boulder-va/main.go:
```diff
@@ -39,18 +39,18 @@ func main() {
-			SimpleHTTPSPort: 443,
-			DVSNIPort:       443,
+			HTTPPort:  80,
+			HTTPSPort: 443,
```
why would you need it if you say in the earlier thread that this PR implements https://github.com/ietf-wg-acme/acme/pull/7, where SimpleHTTP over TLS is forbidden?

bifurcation replied:
I'm not going to mess with SimpleHTTP in this PR.  It's dead tech.  We can pull the HTTPSPort when we kill SimpleHTTP.

kuba commented on va/validation-authority.go:
```diff
@@ -200,29 +201,16 @@ func (va *ValidationAuthorityImpl) resolveAndConstructDialer(name, defaultPort s
-	if input.TLS == nil || (input.TLS != nil && *input.TLS) {
+	scheme := "http"
+	port := va.httpPort
+	if useTLS {
```
https://github.com/ietf-wg-acme/acme/pull/7 disallows TLS for simplehttp

kuba commented on cmd/boulder-va/main.go:
```diff
@@ -39,18 +39,18 @@ func main() {
 		}
-		if c.VA.PortConfig.SimpleHTTPSPort != 0 {
-			pc.SimpleHTTPSPort = c.VA.PortConfig.SimpleHTTPSPort
+		if c.VA.PortConfig.HTTPSPort != 0 {
```
ditto, https://github.com/letsencrypt/boulder/pull/774/files#r41309836

kuba commented on test/boulder-config.json:
```diff
@@ -134,9 +134,9 @@
-      "simpleHTTPSPort": 5001,
-      "dvsniPort": 5001
+      "httpPort": 5001,
+      "httpsPort": 5001,
```
ditto, https://github.com/letsencrypt/boulder/pull/774/files#r41309836

kuba commented on test/boulder-config.json:
```diff
@@ -134,9 +134,9 @@
-      "dvsniPort": 5001
+      "httpPort": 5001,
+      "httpsPort": 5001,
+      "tlsPort": 5001
```
`httpPort` and `tlsPort` might not be the same if the client wants to test both challenges - it's impossible to operate socket on the same port, both for TLS and non-TLS connections. Please use different ports, otherwise integration testing will be incredibly difficult.

bifurcation replied:
Separate PR.  This is just renaming.

kuba replied:
it makes client-server integration testing hard; can you update it now, so that I can work on the client implementation?

jmhodges replied:
A separate PR is fine. We've got enough going on in this one.

kuba commented on va/validation-authority.go:
```diff
@@ -42,24 +42,24 @@ var ErrTooManyCNAME = errors.New("too many CNAME/DNAME lookups")
+	DNSResolver  core.DNSResolver
+	IssuerDomain string
+	httpPort     int
+	httpsPort    int
```
ditto, https://github.com/letsencrypt/boulder/pull/774/files#r41309836

kuba commented on va/validation-authority.go:
```diff
@@ -42,24 +42,24 @@ var ErrTooManyCNAME = errors.New("too many CNAME/DNAME lookups")
-	SimpleHTTPSPort int
-	DVSNIPort       int
+	HTTPPort  int
+	HTTPSPort int
```
ditto, https://github.com/letsencrypt/boulder/pull/774/files#r41309836

kuba commented on va/validation-authority.go:
```diff
@@ -67,12 +67,12 @@ func NewValidationAuthorityImpl(pc *PortConfig, stats statsd.Statter, clk clock.
-		clk:             clk,
+		log:       logger,
+		httpPort:  pc.HTTPPort,
+		httpsPort: pc.HTTPSPort,
```
https://github.com/letsencrypt/boulder/pull/774/files#r41309836

kuba commented on va/validation-authority_test.go:
```diff
@@ -246,18 +244,19 @@ func TestSimpleHttpTLS(t *testing.T) {
 	test.AssertNotError(t, err, "failed to get test server port")
 	stats, _ := statsd.NewNoopClient()
-	va := NewValidationAuthorityImpl(&PortConfig{SimpleHTTPSPort: port}, stats, clock.Default())
+	va := NewValidationAuthorityImpl(&PortConfig{HTTPSPort: port}, stats, clock.Default())
```
again, https://github.com/ietf-wg-acme/acme/pull/7 disallows TLS and this entire test function shall be removed

jmhodges commented on va/validation-authority.go:
```diff
@@ -330,18 +318,107 @@ func (va *ValidationAuthorityImpl) validateSimpleHTTP(identifier core.AcmeIdenti
+	}
+
+	// Make a connection with SNI = nonceName
+	portString := fmt.Sprintf("%d", va.tlsPort)
```
strconv.Itoa()

jmhodges commented:
Those little nits are all!

jmhodges commented:
LGTM

pde commented:
I think #903 should probably have been merged into this branch before it landed.  We now have a non-spec-compliant TLS-SNI-01 in deployment :(
